File: C_plus_plus/Sort_Colors.cpp
class Solution {
public:
    void sortColors(int A[], int n) {
        int r = 0, w = 0, b = n-1,tmp;
        for(int i = 0 ; i < n ; ++i){
            if(w+r+n-1-b==n) break;
            if(A[i]==0) A[r++] = 0;
            else if(A[i]==2){
                tmp = A[b];
                A[b--] = 2;
                A[i--] = tmp; 
            }
            else if(A[i]==1) ++w;
        }
        for(int i = r ; i <= b ; ++i)
            A[i] = 1;
    }
};

File: java/Sort_Colors.java
// One pass solution with constant space
// Pass the array , when we come cross a 0, we swap it with the first one that we haven't set a 0 to it.
// when we come cross a 2, we swap it with the last one that we haven't set a 2 to it.

public class Solution {
    public void sortColors(int[] nums) {
        int low = 0, high = nums.length-1;
        for (int i = 0; i <= high; ++i) {
            if (nums[i] == 0) {
                nums[i] = nums[low]; 
                nums[low++] = 0;
            }
            else if (nums[i] == 2) {
                nums[i--] = nums[high];
                nums[high--] = 2;
            }
        }
    }
}
